### DISCUSSION TITLE
Using mozjpeg in place of libjpeg-turbo

### DISCUSSION THREAD

**Discussion by quentin-bettoum** [libvips/libvips#3360] [Help wanted]
Hello,

I wanted to use `mozjpeg` to use the related arguments of `jpegsave` such as `trellis_quant`, `optimize_scans` and `quant_table`.

I first thought it could be a matter of installing mozjpeg in place of libjpeg-turbo, but even after doing this, the below command would still output some warnings.
```bash
vips copy forest.jpg forest-opti.jpg[Q=30,strip,interlace,optimize-coding,optimize-scans,trellis-qu
ant,quant-table=3]

(vips:21581): VIPS-WARNING **: 14:50:18.119: ignoring trellis_quant

(vips:21581): VIPS-WARNING **: 14:50:18.119: ignoring optimize_scans

(vips:21581): VIPS-WARNING **: 14:50:18.119: ignoring quant_table
```

This is strange because the output image is different from the one produced when `libjpeg-turbo` is installed, which let me think that mozjpeg might be used, but I don't understand why these options are still ignored.

While searching, I've come across [this Dockerfile](https://github.com/jcupitt/docker-builds/blob/master/libvips-mozjpeg-ubuntu20.04/Dockerfile) which was mentioned here and there.
I am wondering if `libvips` should be compiled following the dockerfile in order to use `mozjpeg`?
Or maybe I need to install `mozjpeg` in a different way?

I am on Linux (arch) and I use these versions:
[`vips-8.14.1`](https://archlinux.org/packages/community/x86_64/libvips/) which come from the official Arch repository. They are using [this PKGBUILD](https://github.com/archlinux/svntogit-community/blob/packages/libvips/trunk/PKGBUILD).
[`mozjpeg 4.1.1-2`](https://aur.archlinux.org/packages/mozjpeg) from the Arch User Repository which is using [this PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=mozjpeg).

Thanks for reading.

**Comment by jcupitt** [comment]
Hi @quentin-bettoum,

It's surprisingly difficult to use mozjpeg. The jpeg library is really fundamental, so swapping it out means changing a huge number of packages. 

libjpeg-turbo manages this by being very careful about binary compatibility, so you can substitute the libjpeg-turbo shared library for the regular libjpeg one and everything still works. However, mozjpeg is *not* binary compatible, so you can't do this. Instead, you must rebuild everything. The dockerfiles that use mozjpeg first build mozjpeg itself, and then build custom versions of all direct and indirect libvips dependencies which use libjpeg, and then finally build libvips.

There has been talk for a long time of unifying mozjpeg and libjpeg-turbo to make this pain go away, but I don't know if it's happened yet, I've not checked for a while.

**Reply by quentin-bettoum** [reply]
Thanks @jcupitt

It seems that the [sharp module](https://github.com/lovell/sharp) is using a build of libvips which includes mozjpeg. Apparently, the homebrew package manager also seems to build libvips with mozjpeg. [Here is their build file in ruby](https://github.com/Homebrew/homebrew-core/blob/955021b2c365f4d020968962f61f5f10aea1b120/Formula/vips.rb) if it can help anyone.

From what I understand, Arch Linux builds libvips without mozjpeg.
I think I will try to investigate a bit further to see if I can take inspiration from the above-mentioned build and make it work on my system.

**Comment by akash-akya** [comment]
Cases like these are exactly where Nix shines. With nix, you can just build a vips with mozjpeg without affecting any of the system packages.
Just now, I copied [nix vips pkg](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/graphics/vips/default.nix) and swapped libjpeg dependency with moz and few minor changes, and it just worked! 

If you feel adventurous, you can [install nix](https://nixos.org/download.html) and give it a shot!

`default.nix`
```nix
with (import <nixpkgs> {});
let
in stdenv.mkDerivation rec {
  pname = "vips";
  version = "8.14.1";

  outputs = [ "bin" "out" "man" "dev" ] ++ lib.optionals (!stdenv.isDarwin) [ "devdoc" ];

  src = fetchFromGitHub {
    owner = "libvips";
    repo = "libvips";
    rev = "v${version}";
    hash = "sha256-ajGVSVjnv78S/Xd3Aqn0N87I7m39DWKZHAQjwbog+5U=";
    # Remove unicode file names which leads to different checksums on HFS+
    # vs. other filesystems because of unicode normalisation.
    postFetch = ''
      rm -r $out/test/test-suite/images/
    '';
  };

  nativeBuildInputs = [
    pkg-config
    meson
    ninja
    docbook-xsl-nons
    gobject-introspection
  ] ++ lib.optionals (!stdenv.isDarwin) [
    gtk-doc
  ];

  buildInputs = [
    glib
    libxml2
    expat
    (python3.withPackages (p: [ p.pycairo ]))
    # Optional dependencies
    mozjpeg
    libexif
    librsvg
    poppler
    libgsf
    libtiff
    fftw
    lcms2
    libpng
    libimagequant
    imagemagick
    pango
    orc
    matio
    cfitsio
    libwebp
    openexr
    openjpeg
    libjxl
    openslide
    libheif
    which
  ] ++ lib.optionals stdenv.isDarwin [ ];

  # Required by .pc file
  propagatedBuildInputs = [
    glib
  ];

  mesonFlags = [
    "-Dcgif=disabled"
    "-Dspng=disabled"
    "-Dpdfium=disabled"
    "-Dnifti=disabled"
  ] ++ lib.optionals (!stdenv.isDarwin) [
    "-Dgtk_doc=true"
  ];

  meta = with lib; {
    changelog = "https://github.com/libvips/libvips/blob/${src.rev}/ChangeLog";
    homepage = "https://libvips.github.io/libvips/";
    description = "Image processing system for large images";
    license = licenses.lgpl2Plus;
    maintainers = with maintainers; [ kovirobi ];
    platforms = platforms.unix;
  };
}
```

Build Package
```
$ nix-build default.nix
```

```shell
$ otool -L ./result-bin/bin/vips | grep libvips
	/nix/store/li1rss6nj8dfiks6z50pz2h8b9lhx7h3-vips-8.14.1/lib/libvips.42.dylib (compatibility version 59.0.0, current version 59.1.0)

$ otool -L /nix/store/li1rss6nj8dfiks6z50pz2h8b9lhx7h3-vips-8.14.1/lib/libvips.42.dylib | grep jpeg
	/nix/store/l0m2n5nvsw0chi95hxp1hvc7bczzjylh-mozjpeg-4.0.3/lib/libjpeg.62.dylib (compatibility version 62.0.0, current version 62.3.0)
	/nix/store/d367shx1vwbm7nzwrjxd47fbgbjr3hiv-openjpeg-2.4.0/lib/libopenjp2.7.dylib (compatibility version 7.0.0, current version 2.4.0)
```

**Reply by quentin-bettoum** [reply]
Thanks @akash-akya 

Nix was on my "things to try" list, and I shall definitely give it a try.
